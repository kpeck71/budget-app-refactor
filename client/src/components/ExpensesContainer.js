import React, { Component } from 'react';
import axios from 'axios'
import Expense from './Expense'
import NewExpenseForm from './NewExpenseForm'
import EditExpenseForm from './EditExpenseForm'

class ExpensesContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      expenses: [],
      editingExpenseId: null
    }
    this.addNewExpense = this.addNewExpense.bind(this)
    this.removeExpense = this.removeExpense.bind(this)
    this.editingExpense = this.editingExpense.bind(this)
    this.editExpense = this.editExpense.bind(this)
  }

  componentDidMount() {
    axios.get('http://localhost:3001/api/v1/expenses.json')
    .then(response => {
      console.log(response)
      this.setState({
        expenses: response.data
      })
    })
    .catch(error => console.log(error))
  }

  addNewExpense(title, amount, category) {
     axios.post( '/api/v1/expenses', { expense: {title, amount, category} })
     .then(response => {
         console.log(response)
         const expenses = [ ...this.state.expenses, response.data ]
         this.setState({expenses})
     })
     .catch(error => {
         console.log(error)
     })
   }

   removeExpense(id) {
      axios.delete( '/api/v1/expenses/' + id )
      .then(response => {
          const expenses = this.state.expenses.filter(
              expense => expense.id !== id
          )
          this.setState({expenses})
      })
      .catch(error => console.log(error))
    }

    editingExpense(id) {
      this.setState({
        editingExpenseId: id
      })
    }

    editExpense(id, title, amount, category) {
      axios.put( '/api/v1/expenses/' + id, {
          expense: {
              title,
              amount,
              category
          }
      })
      .then(response => {
          console.log(response);
          const expenses = this.state.expenses;
          expenses[id-1] = {id, title, amount, category}
          this.setState(() => ({
              expenses,
              editingExpenseId: null
          }))
      })
      .catch(error => console.log(error));
    }

  render() {
    return (
      <div className="expenses-container container-fluid">
        <div className="row">
          <h1>Expenses</h1>
        {this.state.expenses.map( expense => {
          if (this.state.editingExpenseId === expense.id ) {
            return (<EditExpenseForm
                      expense={expense}
                      key={expense.id}
                      editExpense={this.editExpense}
                    />)
          } else {
          return (<Expense
                    expense={expense}
                    key={expense.id}
                    onRemoveExpense={this.removeExpense}
                    editingExpense={this.editingExpense}
                    /> )
          }
        })}
        </div>
        <NewExpenseForm onNewExpense={this.addNewExpense} />

      </div>
      )
    }
  }

export default ExpensesContainer;
