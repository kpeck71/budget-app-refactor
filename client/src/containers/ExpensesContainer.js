import React, { Component } from 'react';
import axios from 'axios';
import Expense from '../components/Expense';
import NewExpenseForm from '../components/NewExpenseForm';
import EditExpenseForm from '../components/EditExpenseForm';

class ExpensesContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      expenses: [],
      budgets: [],
      editingExpenseId: null,
      isHidden: true
    }
    this.addNewExpense = this.addNewExpense.bind(this)
    this.removeExpense = this.removeExpense.bind(this)
    this.editingExpense = this.editingExpense.bind(this)
    this.editExpense = this.editExpense.bind(this)
  }
  toggleHidden() {
    this.setState({
      isHidden: !this.state.isHidden
    })
  }

  componentDidMount() {
    axios.get('http://localhost:3001/api/v1/expenses.json')
    .then(response => {
      // console.log(response)
      this.setState({
        expenses: response.data
      })
    })
    .catch(error => console.log(error))

    axios.get('http://localhost:3001/api/v1/budgets.json')
    .then(response => {
      this.setState({
        budgets: response.data
      })
    })
    .catch(error => console.log(error))
  }

  addNewExpense(title, amount, category) {
     axios.post( '/api/v1/expenses', { expense: {title, amount, category} })
     .then(response => {
         console.log(response)
         const expenses = [ ...this.state.expenses, response.data ]
         this.setState({expenses})
     })
     .catch(error => {
         console.log(error)
     })
   }

   removeExpense(id) {
      axios.delete( '/api/v1/expenses/' + id )
      .then(response => {
          const expenses = this.state.expenses.filter(
              expense => expense.id !== id
          )
          this.setState({expenses})
      })
      .catch(error => console.log(error))
    }

    editingExpense(id) {
      this.setState({
        editingExpenseId: id
      })
    }

    editExpense(id, title, amount, category) {
      axios.put( '/api/v1/expenses/' + id, {
          expense: {
              title,
              amount,
              category
          }
      })
      .then(response => {
          console.log(response);
          const expenses = this.state.expenses;
          expenses[id-1] = {id, title, amount, category}
          this.setState(() => ({
              expenses,
              editingExpenseId: null
          }))
      })
      .catch(error => console.log(error));
    }

    calculateBudget() {
      let expenseTotal = this.state.expenses.reduce((acc, cv) => acc + parseInt(cv.amount), 0)
      let incomeTotal = this.state.budgets.reduce((acc, cv) => acc + cv.amount, 0)
      return incomeTotal - expenseTotal
    }

  render() {
    const expenseClassName = this.calculateBudget() <= 500 ? 'expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert' : 'expense-ok col-md-auto border rounded p-2 m-2 mx-5 border-primary'

    return (

      <div className="expenses-container container-fluid">

        <div className={expenseClassName}>
          You have this much left to spend: ${this.calculateBudget()}
        </div>
        {!this.state.isHidden && <NewExpenseForm onNewExpense={this.addNewExpense} /> }
        <button className="btn btn-outline-warning" onClick={this.toggleHidden.bind(this)} type="submit">New Expense</button>
        <div className="row justify-content-md-center">
          {this.state.expenses.map( expense => {
            if (this.state.editingExpenseId === expense.id ) {
              return (<EditExpenseForm
                        expense={expense}
                        key={expense.id}
                        editExpense={this.editExpense}
                      />)
            } else {
            return (<Expense
                      expense={expense}
                      key={expense.id}
                      onRemoveExpense={this.removeExpense}
                      editingExpense={this.editingExpense}
                      /> )
            }
          })}
        </div>
        <br/>
      </div>
      )
    }
  }

export default ExpensesContainer;
