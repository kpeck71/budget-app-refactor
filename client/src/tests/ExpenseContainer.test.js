import { shallow } from "enzyme";
import React from "react";
import ReactDOM from "react-dom";
import ExpensesContainer from "../containers/ExpensesContainer"
import './enzyme_setup'

describe("ExpensesContainer", () => {
  it("should render even with no expenses", () => {
    const expenses = []
    const wrapper = shallow(
      <ExpensesContainer expenses={expenses} />
    );
    expect(wrapper.find(".expenses-container").exists()).toBe(true);
  });
})

describe("when given `expenses`", () => {
  const expenses = [
    { id: 1, title: "Rent", amount: 2000, category: "essentials" },
    { id: 2, title: "MetroCard", amount: 121, category: "essentials"}
  ]

  it("should render correct number of expenses", () => {
    const wrapper = shallow(
      <ExpensesContainer expenses={expenses} />
    );
    expect(wrapper.find("title")).toMatchSnapshot();
    // expect(wrapper.find(".single-expense").length).toEqual(2)
  });
})
