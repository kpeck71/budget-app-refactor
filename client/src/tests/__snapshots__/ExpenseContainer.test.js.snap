// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when given \`expenses\` should render correct number of expenses 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): ShallowWrapper {
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <ExpensesContainer
      expenses={
        Array [
          Object {
            "amount": 2000,
            "category": "essentials",
            "id": 1,
            "title": "Rent",
          },
          Object {
            "amount": 121,
            "category": "essentials",
            "id": 2,
            "title": "MetroCard",
          },
        ]
      }
    />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateError": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert"
          >
            You have this much left to spend: $
            0
          </div>,
          false,
          <button
            className="btn btn-outline-warning"
            onClick={[Function]}
            type="submit"
          >
            New Expense
          </button>,
          <div
            className="row justify-content-md-center"
          />,
          <br />,
        ],
        "className": "expenses-container container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "You have this much left to spend: $",
              0,
            ],
            "className": "expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert",
          },
          "ref": null,
          "rendered": Array [
            "You have this much left to spend: $",
            0,
          ],
          "type": "div",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "New Expense",
            "className": "btn btn-outline-warning",
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": "New Expense",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "row justify-content-md-center",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
      ],
      "type": "div",
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert"
            >
              You have this much left to spend: $
              0
            </div>,
            false,
            <button
              className="btn btn-outline-warning"
              onClick={[Function]}
              type="submit"
            >
              New Expense
            </button>,
            <div
              className="row justify-content-md-center"
            />,
            <br />,
          ],
          "className": "expenses-container container-fluid",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "You have this much left to spend: $",
                0,
              ],
              "className": "expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert",
            },
            "ref": null,
            "rendered": Array [
              "You have this much left to spend: $",
              0,
            ],
            "type": "div",
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "New Expense",
              "className": "btn btn-outline-warning",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "New Expense",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "row justify-content-md-center",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
        ],
        "type": "div",
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
          "lifecycles": Object {
            "componentDidUpdate": Object {
              "onSetState": true,
            },
            "getDerivedStateFromProps": true,
            "getSnapshotBeforeUpdate": true,
            "setState": Object {
              "skipsComponentDidUpdateOnNullish": true,
            },
          },
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__rootNodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert"
          >
            You have this much left to spend: $
            0
          </div>,
          false,
          <button
            className="btn btn-outline-warning"
            onClick={[Function]}
            type="submit"
          >
            New Expense
          </button>,
          <div
            className="row justify-content-md-center"
          />,
          <br />,
        ],
        "className": "expenses-container container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "You have this much left to spend: $",
              0,
            ],
            "className": "expense-low col-md-auto border rounded p-2 m-2 mx-5 border-alert",
          },
          "ref": null,
          "rendered": Array [
            "You have this much left to spend: $",
            0,
          ],
          "type": "div",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "New Expense",
            "className": "btn btn-outline-warning",
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": "New Expense",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "row justify-content-md-center",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
      ],
      "type": "div",
    },
  ],
}
`;
